File: app/src/main/java/com/orange/tpms/ue/frag/Frag_id_copy_original.java
package com.orange.tpms.ue.frag;

import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.text.TextUtils;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.TextView;

import com.de.rocket.Rocket;
import com.de.rocket.ue.frag.RoFragment;
import com.de.rocket.ue.injector.Event;
import com.orange.tpms.R;
import com.orange.tpms.adapter.IDCopyAdapter;
import com.orange.tpms.bean.*;
import com.orange.tpms.helper.CopyIDHelper;
import com.orange.tpms.lib.hardware.HardwareApp;
import com.orange.tpms.ue.activity.MainActivity;
import com.orange.tpms.utils.NumberUtil;
import com.orange.tpms.utils.VibMediaUtil;
import com.orange.tpms.widget.LoadingWidget;
import com.orange.tpms.widget.ScanWidget;
import com.orange.tpms.widget.SensorWayWidget;
import com.orange.tpms.widget.TitleWidget;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import com.de.rocket.ue.injector.BindView;

import bean.mmy.MMyBean;

/**
 * ID COPY 详情页
 * Created by haide.yin() on 2019/4/4 9:28.
 */
public class Frag_id_copy_original extends Frag_base {
    String ObdHex="00";
    @BindView(R.id.v_title_bar)
    TitleWidget twTitle;//Title
    @BindView(R.id.rv_id_copy)
    RecyclerView rvIDCopy;//IDCopy
    @BindView(R.id.tv_content)
    TextView tvContent;//title
    @BindView(R.id.sww_select)
    SensorWayWidget swwSelect;
    @BindView(R.id.ldw_loading)
    LoadingWidget lwLoading;//Loading
//    @BindView(R.id.scw_tips)
//    ScanWidget scwTips;//Tips

    private IDCopyAdapter idCopyAdapter;//适配器
    private LinearLayoutManager linearLayoutManager;//列表表格布局
    private CopyIDHelper copyIDHelper;
    private HardwareApp.DataReceiver dataReceiver;
    private VibMediaUtil vibMediaUtil;//音效与振动

    @Override
    public int onInflateLayout() {
        return R.layout.frag_id_copy_original;
    }

    @Override
    public void initViewFinish(View inflateView) {
        initView();
        initHelper();
    }

    @Override
    public void onDestroyView(){//与initViewFinish对应的生命周期
        super.onDestroyView();
        vibMediaUtil.release();
    }

    @Override
    public void onNexts(Object o) {
//        ObdHex=((MainActivity)activity).itemDAO.GetHex(PublicBean.SelectMake, PublicBean.SelectModel,PublicBean.SelectYear);
        tvContent.setText(ObdHex);
    }

    @Override
    public void onKeyTrigger(){
        if(swwSelect.isShown()){
            swwSelect.pllTrigger.performClick();
        }

        //避免回调两次出现回退异常,判断当前页面在栈顶才会读取
        if(Rocket.isTopRocketStack(Frag_id_copy_original.class.getSimpleName())){
            if (swwSelect.getSelectType() == SensorWayWidget.SELECT_TYPE_TRIGGER) {
                copyIDHelper.trigger(ObdHex);
            }
        }
    }

    @Override
    public void onKeyScan(){
        Rocket.writeOuterLog("Frag_id_copy_original::onKeyScan");
        if(swwSelect.isShown()){
            swwSelect.pllScan.performClick();
        }

        //避免回调两次出现回退异常,判断当前页面在栈顶才会读取
        if(Rocket.isTopRocketStack(Frag_id_copy_original.class.getSimpleName())){
            Rocket.writeOuterLog("Frag_id_copy_original::onKeyScan->isTopRocketStack");
            if (swwSelect.getSelectType() == SensorWayWidget.SELECT_TYPE_SCAN) {
                Rocket.writeOuterLog("Frag_id_copy_original::onKeyScan->SelectType:SELECT_TYPE_SCAN");
                HardwareApp.getInstance().scan();
                lwLoading.show(getResources().getString(R.string.app_scaning));
            }else{
                Rocket.writeOuterLog("Frag_id_copy_original::onKeyScan->SelectType:"+swwSelect.getSelectType());
            }
        }
    }

    @Override
    public boolean onBackPresss(){
        return true;
    }

    @Event(R.id.bt_new_sensor_list)
    private void newSensor(View view){
        vibMediaUtil.playVibrate();
        if(checkHasSensor()){
            if(!haveSameSensorid()){
                //主动释放,不然会因为生命周期的先后顺序导致注册解绑冲突
                HardwareApp.getInstance().switchScan(false);
                HardwareApp.getInstance().removeDataReceiver(dataReceiver);
                //传递参数
//                IDCopyFragBean idCopyFragBean = new IDCopyFragBean();
//                idCopyFragBean.setOrignalSendorid(getSensoridList());
//                idCopyFragBean.setmMyBean(mMyBean);
                PublicBean.SensorList=getSensoridList();
                toFrag(Frag_id_copy_new.class, true,true,"");
            }else{
                toast(R.string.app_sensor_repeated);
            }
        }else{
            toast(R.string.app_no_sensor_set);
        }
    }

    @Event(R.id.bt_menue)
    private void menue(View view){
        toFrag(Frag_home.class,true,true,null,true);
    }

    /**
     * 初始化页面
     */
    private void initView() {
        swwSelect.setTitle(getResources().getString(R.string.app_original_sensor));
        //音效与震动
        vibMediaUtil = new VibMediaUtil(activity);
        //设置标题
        twTitle.setTvTitle(R.string.app_copy_original);
        //返回
        twTitle.setOnBackListener((view) -> back());
        //配置RecyclerView,每行是哪个元素
        if (linearLayoutManager == null) {
            linearLayoutManager = new LinearLayoutManager(activity);
        }
        rvIDCopy.setLayoutManager(linearLayoutManager);
        idCopyAdapter = new IDCopyAdapter(activity);
        rvIDCopy.setAdapter(idCopyAdapter);
        //数据源
        List<IDCopyBean> numberList = new ArrayList<>();
        IDCopyBean titleBean = new IDCopyBean("",getString(R.string.app_id_clear),getString(R.string.app_psi),getString(R.string.app_temp),getString(R.string.app_bat_clear),false);
        IDCopyBean frBean = new IDCopyBean("FR","","","","",false);
        IDCopyBean rrBean = new IDCopyBean("RR","","","","",false);
        IDCopyBean rlBean = new IDCopyBean("RL","","","","",false);
        IDCopyBean flBean = new IDCopyBean("FL","","","","",false);
        numberList.add(titleBean);
        numberList.add(frBean);
        numberList.add(rrBean);
        numberList.add(rlBean);
        numberList.add(flBean);
        idCopyAdapter.setItems(numberList);
        idCopyAdapter.notifyDataSetChanged();
        //选择方式
        swwSelect.setOnKeyinClickListener(() -> updateEditable());
        //硬件
        HardwareApp.getInstance().switchScan(true);
        dataReceiver = new HardwareApp.DataReceiver() {
            @Override
            public void scanReceive() {

            }

            @Override
            public void scanMsgReceive(String content) {
                Rocket.writeOuterLog("Frag_id_copy_original::scanMsgReceive->content:"+content);
                lwLoading.hide();
                if(!content.contains(":") && !content.contains("*")){
                    if(!content.equals("nofound")){
                        toast(R.string.app_invalid_sensor_qrcode);
                    }else{
                        toast(R.string.app_scan_code_timeout);
                    }
                    return;
                }
                String sensorid;
                if(content.contains("**")){
                    sensorid = MMYQrCodeBean.toQRcodeBean(content).getMmyNumber();
                }else{
                    sensorid = SensorQrCodeBean.toQRcodeBean(content).getSensorID();
                }
                if(TextUtils.isEmpty(sensorid)){
                    toast(R.string.app_invalid_sensor_qrcode);
                    return;
                }
                Rocket.writeOuterLog("Frag_id_copy_new::scanMsgReceive->sensorid:"+sensorid);
                vibMediaUtil.playBeep();
                if(!haveSameSensorid(sensorid)){
                    updateSensorid(sensorid);
                }else{
                    toast(R.string.app_sensor_repeated);
                }
            }

            @Override
            public void uart2MsgReceive(String content) {

            }
        };
        HardwareApp.getInstance().addDataReceiver(dataReceiver);
    }

    /**
     * 插入传感器id
     */
    private void updateSensorid(String sensorid) {
        for (int i = 1; i < idCopyAdapter.getItems().size(); i++) {
            IDCopyBean idCopyBean = idCopyAdapter.getItems().get(i);
            //sensorid为空才插入
            if(TextUtils.isEmpty(idCopyBean.getSensorid())){
                idCopyBean.setSensorid(sensorid);
                idCopyAdapter.setItem(i,idCopyBean);
                rvIDCopy.setAdapter(idCopyAdapter);
                checkSelectFinish();
                return;
            }
        }
    }

    /**
     * 插入传感器id
     */
    private void updateSensorid(String sensorid,String psi,String temp,String bat) {
        for (int i = 1; i < idCopyAdapter.getItems().size(); i++) {
            IDCopyBean idCopyBean = idCopyAdapter.getItems().get(i);
            //sensorid为空才插入
            if(TextUtils.isEmpty(idCopyBean.getSensorid())){
                idCopyBean.setSensorid(sensorid);
                idCopyBean.setPsi(psi);
                idCopyBean.setTemp(temp);
                idCopyBean.setBat(bat);
                idCopyAdapter.setItem(i,idCopyBean);
                rvIDCopy.setAdapter(idCopyAdapter);
                checkSelectFinish();
                return;
            }
        }
    }

    /**
     * 刷新是否能够编辑的状态
     */
    private void updateEditable() {
        for (int i = 1; i < idCopyAdapter.getItems().size(); i++) {
            IDCopyBean idCopyBean = idCopyAdapter.getItems().get(i);
            idCopyBean.setEditable(true);
            idCopyAdapter.setItem(i,idCopyBean);
        }
        rvIDCopy.setAdapter(idCopyAdapter);
    }

    /**
     * 检测是否完成选择
     */
    private boolean checkSelectFinish() {
        boolean finish = true;
        for (int i = 1; i < idCopyAdapter.getItems().size(); i++) {
            IDCopyBean idCopyBean = idCopyAdapter.getItems().get(i);
            //sensorid都不为空才行
            if(TextUtils.isEmpty(idCopyBean.getSensorid())){
                finish = false;
            }
        }
        return finish;
    }

    /**
     * 获取传感器id列表
     */
    private List<String> getSensoridList() {
        List<String> sensorList = new ArrayList<>();
        for (int i = 1; i < idCopyAdapter.getItems().size(); i++) {
            IDCopyBean idCopyBean = idCopyAdapter.getItems().get(i);
            //sensorid都不为空才行
            if(!TextUtils.isEmpty(idCopyBean.getSensorid())){
                sensorList.add(idCopyBean.getSensorid());
            }
        }
        return sensorList;
    }

    /**
     * 检测是否有数据了
     */
    private boolean checkHasSensor() {
        for (int i = 1; i < idCopyAdapter.getItems().size(); i++) {
            IDCopyBean idCopyBean = idCopyAdapter.getItems().get(i);
            //只要有一个不为空即可
            if(!TextUtils.isEmpty(idCopyBean.getSensorid())){
                return true;
            }
        }
        return false;
    }

    /**
     * 检测是否有重复的
     */
    private boolean haveSameSensorid(String sensorid) {
        for (int i = 1; i < idCopyAdapter.getItems().size(); i++) {
            IDCopyBean idCopyBean = idCopyAdapter.getItems().get(i);
            if(!TextUtils.isEmpty(sensorid) && !TextUtils.isEmpty(idCopyBean.getSensorid())){
                if(sensorid.equals(idCopyBean.getSensorid())){
                    return true;
                }
            }
        }
        return false;
    }

    /**
     * 检测是否有重复的
     */
    private boolean haveSameSensorid() {
        Set<String> set = new HashSet<>();
        Set<String> exist = new HashSet<>();
        for (IDCopyBean idCopyBean : idCopyAdapter.getItems()) {
            String sensorid = idCopyBean.getSensorid();
            if(!TextUtils.isEmpty(sensorid)){
                if (set.contains(sensorid)) {
                    exist.add(sensorid);
                } else {
                    set.add(sensorid);
                }
            }
        }
        return !exist.isEmpty();
    }

    /**
     * 初始化Helper
     */
    private void initHelper() {
        copyIDHelper = new CopyIDHelper();
        //开始请求
        copyIDHelper.setOnPreRequestListener(() -> lwLoading.show(getResources().getString(R.string.app_data_reading)));
        //结束请求
        copyIDHelper.setOnFinishRequestListener(() -> lwLoading.hide());
        //请求失败
        copyIDHelper.setOnFailedRequestListener((object -> toast(object.toString(), 2000)));
        //读取传感器情况
        copyIDHelper.setOnStrggerSuccessListener((sensorid, psi, temp, bat) -> {
            vibMediaUtil.playBeep();
            if(!haveSameSensorid(sensorid)){
                updateSensorid(sensorid,psi,temp,bat);
            }else{
                toast(R.string.app_sensor_repeated);
            }
        });
        //进度
        copyIDHelper.setOnProgressListener(progress -> {
            if (lwLoading.getVisibility() != View.VISIBLE) {
                lwLoading.show();
            }
            String content = NumberUtil.toFormate(progress);
            lwLoading.getTvLoading().setText(String.valueOf(content + "%"));
        });
    }
}

File: app/src/main/java/com/orange/tpms/ue/kt_frag/Frag_Idcopy_original.kt
package com.orange.tpms.ue.kt_frag


import android.app.Dialog
import android.text.TextUtils
import android.view.KeyEvent
import android.view.View
import android.widget.TextView
import androidx.fragment.app.Fragment
import com.orange.jzchi.jzframework.JzActivity
import com.orange.jzchi.jzframework.callback.SetupDialog
import com.orange.tpms.R
import com.orange.tpms.RootFragement
import com.orange.tpms.adapter.IDCopyAdapter
import com.orange.tpms.bean.IDCopyBean
import com.orange.tpms.bean.MMYQrCodeBean
import com.orange.tpms.bean.PublicBean
import com.orange.tpms.lib.hardware.HardwareApp
import com.orange.tpms.ue.activity.KtActivity
import com.orange.tpms.ue.dialog.EmptyDialog
import com.orange.tpms.ue.dialog.SensorWay
import com.orange.tpms.utils.OgCommand
import com.orange.tpms.utils.VibMediaUtil
import kotlinx.android.synthetic.main.data_loading.*
import kotlinx.android.synthetic.main.fragment_frag__idcopy_original.view.*
import java.util.*


/**
 * A simple [Fragment] subclass.
 *
 */
class Frag_Idcopy_original : RootFragement(R.layout.fragment_frag__idcopy_original) {
    lateinit var rvIDCopy: androidx.recyclerview.widget.RecyclerView//IDCopy
    lateinit var tvContent: TextView//title
    var ObdHex = "00"
    var idcount=8
    lateinit var idCopyAdapter: IDCopyAdapter//适配器
    lateinit var linearLayoutManager: androidx.recyclerview.widget.LinearLayoutManager//列表表格布局
    lateinit var dataReceiver: HardwareApp.DataReceiver
    lateinit var vibMediaUtil: VibMediaUtil//音效与振动
    override fun viewInit() {
        ObdHex=(activity as KtActivity).itemDAO.GetHex(PublicBean.SelectMake,PublicBean.SelectModel,PublicBean.SelectYear)
        rootview.tv_content.text="${PublicBean.SelectMake}/${PublicBean.SelectModel}/${PublicBean.SelectYear}"
        rootview.bt_menue.setOnClickListener { Trigger() }
        rvIDCopy=rootview.findViewById(R.id.rv_id_copy)
        tvContent=rootview.findViewById(R.id.tv_content)
        initView()
        rootview.bt_new_sensor_list.setOnClickListener {
            vibMediaUtil.playVibrate()
            if (checkHasSensor()) {
                if (!haveSameSensorid()) {
                    try{
                        HardwareApp.getInstance().switchScan(false)
                        HardwareApp.getInstance().removeDataReceiver(dataReceiver)
                    }catch (e:Exception){e.printStackTrace()}
                    PublicBean.SensorList = getSensoridList()
                    JzActivity.getControlInstance().changeFrag(Frag_Idcopy_New(),R.id.frage,"Frag_Idcopy_New",true)
                } else {
                    JzActivity.getControlInstance().toast(R.string.app_sensor_repeated)
                }
            } else {
                JzActivity.getControlInstance().toast(R.string.app_no_sensor_set)
//                Trigger()
            }
        }
        JzActivity.getControlInstance().showDiaLog(R.layout.sensor_way_dialog,false,false, SensorWay())
        updateEditable(true)
    }

    override fun enter() {
        rootview.bt_new_sensor_list.performClick()
    }
    override fun onKeyScan() {
        super.onKeyScan()
        if(run){return}
        JzActivity.getControlInstance().closeDiaLog()
        HardwareApp.getInstance().scan()
        JzActivity.getControlInstance().showDiaLog(R.layout.data_loading,false,true, object :SetupDialog {
            override fun dismess() {

            }

            override fun keyevent(event: KeyEvent): Boolean {
              return false
            }

            override fun setup(rootview: Dialog) {
                rootview.pass.visibility=View.VISIBLE
                rootview.pass.text=resources.getString(R.string.app_scaning)
            }

        })
        Thread{
            handler.post { JzActivity.getControlInstance().closeDiaLog()}
            Thread.sleep(3000)
            run=false
        }.start()
    }
    override fun onPause() {
        super.onPause()
        JzActivity.getControlInstance().closeDiaLog()
        vibMediaUtil.release()
        try{
            HardwareApp.getInstance().switchScan(false)
            HardwareApp.getInstance().removeDataReceiver(dataReceiver)
        }catch (e:Exception){e.printStackTrace()}
    }
    override fun onKeyTrigger() {
        Trigger()
    }
    fun Trigger(){
        if(run){return}
        run=true
        updateEditable(false)
        EmptyDialog(R.layout.data_loading).show()
        Thread{
            val a = OgCommand.GetId(ObdHex, "00")
            handler.post {
                run = false
                if(!JzActivity.getControlInstance().getNowPageTag().equals("Frag_Idcopy_original")){return@post}
                vibMediaUtil.playBeep()
                JzActivity.getControlInstance().closeDiaLog()
                if(a.success){
                    if (!haveSameSensorid(a.id)) {
                        updateSensorid(a.id,""+a.kpa,""+a.c,""+a.vol);
                    } else {
                        JzActivity.getControlInstance().toast(R.string.app_sensor_repeated)
                    }
                }else{
                    JzActivity.getControlInstance().toast(resources.getString(R.string.app_read_failed))
                }
                updateEditable(true)
            }
        }.start()
    }
    /**
     * 初始化页面
     */
    private fun initView() {
        //音效与震动
        vibMediaUtil = VibMediaUtil(activity)
        //配置RecyclerView,每行是哪个元素
        linearLayoutManager = androidx.recyclerview.widget.LinearLayoutManager(activity)
        rvIDCopy.layoutManager = linearLayoutManager
        idCopyAdapter = IDCopyAdapter(activity,idcount)
        rvIDCopy.adapter = idCopyAdapter
        //数据源
        val numberList = ArrayList<IDCopyBean>()
        val titleBean = IDCopyBean(
            "WH",
            "Original",
            getString(R.string.app_psi),
            getString(R.string.app_temp),
            getString(R.string.app_bat_clear),
            false
        )
        val frBean = IDCopyBean("LF", "", "", "", "", false)
        val rrBean = IDCopyBean("RF", "", "", "", "", false)
        val rlBean = IDCopyBean("RR", "", "", "", "", false)
        val flBean = IDCopyBean("LR", "", "", "", "", false)
        numberList.add(titleBean)
        numberList.add(frBean)
        numberList.add(rrBean)
        numberList.add(rlBean)
        numberList.add(flBean)
        idCopyAdapter.items = numberList
        idCopyAdapter.notifyDataSetChanged()
        //硬件
        HardwareApp.getInstance().switchScan(true)
        dataReceiver = object : HardwareApp.DataReceiver {
            override fun scanReceive() {

            }

            override fun scanMsgReceive(content: String) {
                JzActivity.getControlInstance().closeDiaLog()
                if (!content.contains(":") && !content.contains("*")) {
                    if (content != "nofound") {
                        JzActivity.getControlInstance().toast(R.string.app_invalid_sensor_qrcode)
                    } else {
                        JzActivity.getControlInstance().toast(R.string.app_scan_code_timeout)
                    }
                    return
                }
                var sensorid=""
                if (content.contains("**")) {
                    sensorid = MMYQrCodeBean.toQRcodeBean(content).mmyNumber
                } else {
                    if(content.split(":","*").size>=3){
                        sensorid=content.split(":","*")[1]
                    }
                }
                if (TextUtils.isEmpty(sensorid)) {
                    JzActivity.getControlInstance().toast(R.string.app_invalid_sensor_qrcode)
                    return
                }
                vibMediaUtil.playBeep()
                if (!haveSameSensorid(sensorid)) {
                    updateEditable(false)
                    updateSensorid(sensorid)
                    updateEditable(true)
                } else {
                    JzActivity.getControlInstance().toast(R.string.app_sensor_repeated)
                }
            }

            override fun uart2MsgReceive(content: String) {

            }
        }
        HardwareApp.getInstance().addDataReceiver(dataReceiver)
    }

    /**
     * 插入传感器id
     */
    private fun updateSensorid(sensorid: String) {
        for (i in 1 until idCopyAdapter.items.size) {
            val idCopyBean = idCopyAdapter.items[i]
            //sensorid为空才插入
            if (TextUtils.isEmpty(idCopyBean.sensorid)) {
                idCopyBean.sensorid = sensorid
                idCopyAdapter.setItem(i, idCopyBean)
                rvIDCopy.adapter = idCopyAdapter
                checkSelectFinish()
                return
            }
        }
    }

    /**
     * 插入传感器id
     */
    private fun updateSensorid(sensorid: String, psi: String, temp: String, bat: String) {
        for (i in 1 until idCopyAdapter.items.size) {
            val idCopyBean = idCopyAdapter.items[i]
            //sensorid为空才插入
            if (TextUtils.isEmpty(idCopyBean.sensorid)) {
                idCopyBean.sensorid = sensorid
                idCopyBean.psi = psi
                idCopyBean.temp = temp
                idCopyBean.bat = bat
                idCopyAdapter.setItem(i, idCopyBean)
                rvIDCopy.adapter = idCopyAdapter
                checkSelectFinish()
                return
            }
        }
    }

    /**
     * 刷新是否能够编辑的状态
     */
     fun updateEditable(a:Boolean) {
        for (i in 1 until idCopyAdapter.items.size) {
            val idCopyBean = idCopyAdapter.items[i]
            idCopyBean.isEditable = a
            idCopyAdapter.setItem(i, idCopyBean)
        }
        rvIDCopy.adapter = idCopyAdapter
    }

    /**
     * 检测是否完成选择
     */
    private fun checkSelectFinish(): Boolean {
        var finish = true
        for (i in 1 until idCopyAdapter.items.size) {
            val idCopyBean = idCopyAdapter.items[i]
            //sensorid都不为空才行
            if (TextUtils.isEmpty(idCopyBean.sensorid)) {
                finish = false
            }
        }
        return finish
    }

    /**
     * 获取传感器id列表
     */
    private fun getSensoridList(): List<String> {
        val sensorList = ArrayList<String>()
        for (i in 1 until idCopyAdapter.items.size) {
            val idCopyBean = idCopyAdapter.items[i]
            //sensorid都不为空才行
            if (!TextUtils.isEmpty(idCopyBean.sensorid)) {
                sensorList.add(idCopyBean.sensorid)
            }
        }
        return sensorList
    }

    /**
     * 检测是否有数据了
     */
    private fun checkHasSensor(): Boolean {
        for (i in 1 until idCopyAdapter.items.size) {
            val idCopyBean = idCopyAdapter.items[i]
            //只要有一个不为空即可
            if (!TextUtils.isEmpty(idCopyBean.sensorid)) {
                return true
            }
        }
        return false
    }

    /**
     * 检测是否有重复的
     */
    private fun haveSameSensorid(sensorid: String): Boolean {
        for (i in 1 until idCopyAdapter.items.size) {
            val idCopyBean = idCopyAdapter.items[i]
            if (!TextUtils.isEmpty(sensorid) && !TextUtils.isEmpty(idCopyBean.sensorid)) {
                if (sensorid == idCopyBean.sensorid) {
                    return true
                }
            }
        }
        return false
    }

    /**
     * 检测是否有重复的
     */
    private fun haveSameSensorid(): Boolean {
        val set = HashSet<String>()
        val exist = HashSet<String>()
        for (idCopyBean in idCopyAdapter.items) {
            val sensorid = idCopyBean.sensorid
            if (!TextUtils.isEmpty(sensorid)) {
                if (set.contains(sensorid)) {
                    exist.add(sensorid)
                } else {
                    set.add(sensorid)
                }
            }
        }
        return !exist.isEmpty()
    }


}
